### ISSUE TITLE
Consider requirements for extensibility

### ISSUE THREAD

**Issue by lmoureaux** [hep-statistics-serialization-standard/hep-statistics-serialization-standard#9]
As a specification that aims to last for many years, it should be ensured that HS^3 can be extended to accommodate future needs. The best way of doing this is by defining a clear way forward for future extensions.

Typically the lifetime of a feature is the following:
1. A small group of people assembles and decides that a new feature is needed
2. They implement it privately in the tool they know best, maybe after discussing the implementation with the authors of other tools. At this point files using the new feature start to exist
3. People from a second tool look at the files and try to implement the new feature. They likely encounter some problems and the format needs to be changed
4. At some point, the implementation of the feature becomes stable and gets included in the main standard. At this point every tool is expected to start supporting it
5. Experience is accumulated over the years and a new, better way of doing the same thing is developed. The old feature gets deprecated
6. Finally after many years, support for the old feature is removed

The main requirement to enable this workflow is related to the private implementations that will pop up over time. There are multiple ways of doing this. For instance, XML allows anyone to define new elements and attributes using namespaces, linking them in the root element of the tree. This gives a pointer to users when encountering an unknown element, but is quite verbose. A more pragmatic approach followed by CSS and JavaScript implementations is to use vendor prefixes for private features (for instance `-moz-` for Firefox, `-webkit-` for Safari, ...). In this case the specification states that unknown vendor prefixes should be ignored. C++ goes yet another way, prefixing all standard features with `std::` and forbidding anyone else from modifying it, but letting anyone invent their own namespaces.

I think HS^3 should consider how custom extensions should be implemented, including how tools should behave when an unknown vendor extension is encountered (for instance, should a combination tool be allowed to just copy-paste the definitions into the combined file, or should it reject the input?).

**Comment by cburgard** [comment]
Thank you very much for bringing up this important point!

Indeed, the ability to extend the standard by new features is an integral part of the path to make this actually usable, and we haven't had sufficient discussions on that topic yet to warrant a decision I think.

I personally think that we will probably be best served by using the "vendor prefix" version, since the standard will be supported by a handful relatively powerful and high-profile tools (ROOT, pyhf, ...), and most new innovations are I think expected to happen within one of those tools, rather than by a completely new third party - but please feel free to object!

**Comment by jonas-eschle** [comment]
I agree with the point and I think this was already part of the discussion in the beginning: HS3 standardizes _what we can_, and what we can't (not yet), we don't. This means (and encourages!) that every tool can extend HS3! 
If it adds very useful functionality but isn't compatible anymore with the standard, hell, so be it, if it serves a good purpose! (you don't _need_ to use it, but we should not restrict functionality. Compare to now: no standard at all. So better _some_ than "forced all", which nobody will use in the library, as libraries want to extend and florish).But this then allows to extend HS3 (and the prefix seems a very sensible approach to pragmatically avoid conflicts).

I also imagine this standard just to grow over time, with more things tried out, with more discussions giving a clear path forward.

**Comment by cburgard** [comment]
@jonas-eschle do you have a specific preference which version you would like best?
